src/ows/src/main/java/org/geoserver/ows/Dispatcher.java

```diff
@@ -1525,51 +1515,34 @@ Object parseRequestKVP(Class type, Request request) throws Exception {
         return null;
     }
 
+    /**
+     * To be called once it was determined the request is a POST request with an XML request body;
+     * uses {@link XmlRequestReader} to parse the {@link Request#getInput() request input}.
+     *
+     * <p>This method expects {@code request.} {@link Request#getNamespace() namespace}, {@link
+     * Request#getPostRequestEementName() postRequestElementName}, {@link Request#getVersion()
+     * version}, and {@link Request#getService() service} to be set already.
+     */
     Object parseRequestXML(Object requestBean, BufferedReader input, Request request)
             throws Exception {
         // check for an empty input stream
-        // if (input.available() == 0) {
         if (!input.ready()) {
             return null;
         }
 
-        // create stream parser
-        XmlPullParserFactory factory = XmlPullParserFactory.newInstance();
-        factory.setNamespaceAware(true);
-        factory.setValidating(false);
-
-        // parse root element
-        XmlPullParser parser = factory.newPullParser();
-        // parser.setInput(input, "UTF-8");
-        parser.setInput(input);
-        parser.nextTag();
-
-        String namespace = (parser.getNamespace() != null) ? parser.getNamespace() : "";
-        String element = parser.getName();
-        String version = null;
-        String service = null;
-
-        for (int i = 0; i < parser.getAttributeCount(); i++) {
-            if ("version".equals(parser.getAttributeName(i))) {
-                version = parser.getAttributeValue(i);
-            }
-            if ("service".equals(parser.getAttributeName(i))) {
-                service = parser.getAttributeValue(i);
-            }
-        }
-
-        parser.setInput(null);
-
-        // reset input stream
-        input.reset();
+        String namespace = request.getNamespace();
```

Comment:
This gave me pause at the beginning, but indeed, service (and thus readOpPost) are called before this section of code is reached, so all the the values are already set in the request, as parsed from the XML.